pf/tfbridge/resource_state.go

```diff
@@ -115,15 +115,30 @@ func parseResourceStateFromTF(
 	private []byte,
 ) (*upgradedResourceState, error) {
 	tfType := rh.schema.Type().TerraformType(ctx)
-	v, err := state.Unmarshal(tfType)
-	if err != nil {
-		return nil, err
-	}
-	return &upgradedResourceState{state: &resourceState{
-		TFSchemaVersion: rh.schema.ResourceSchemaVersion(),
-		Value:           v,
+	return parseResourceStateFromTFInner(ctx, tfType, rh.schema.ResourceSchemaVersion(), state, private)
+}
+
+func parseResourceStateFromTFInner(
+	ctx context.Context,
+	resourceTerraformType tftypes.Type,
+	resourceSchemaVersion int64,
+	state *tfprotov6.DynamicValue,
+	private []byte,
+) (*upgradedResourceState, error) {
+	rs := &resourceState{
+		TFSchemaVersion: resourceSchemaVersion,
 		Private:         private,
-	}}, nil
+	}
+	if state == nil {
```

Comment:
Tightening to also handle physical nulls, probably unnecessary but just in case.